databases/orientdb/src/main/java/com/gentics/mesh/graphdb/cluster/MeshOHazelcastPlugin.java

```diff
@@ -0,0 +1,129 @@
+package com.gentics.mesh.graphdb.cluster;
+
+import java.io.FileNotFoundException;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+
+import com.hazelcast.config.Config;
+import com.hazelcast.config.FileSystemXmlConfig;
+import com.hazelcast.core.Hazelcast;
+import com.hazelcast.core.HazelcastInstance;
+import com.hazelcast.core.HazelcastInstanceNotActiveException;
+import com.orientechnologies.common.io.OFileUtils;
+import com.orientechnologies.common.log.OLogManager;
+import com.orientechnologies.common.parser.OSystemVariableResolver;
+import com.orientechnologies.common.util.OCallableNoParamNoReturn;
+import com.orientechnologies.common.util.OCallableUtils;
+import com.orientechnologies.orient.core.OSignalHandler;
+import com.orientechnologies.orient.core.Orient;
+import com.orientechnologies.orient.core.db.ODefaultEmbeddedDatabaseInstanceFactory;
+import com.orientechnologies.orient.server.OServer;
+import com.orientechnologies.orient.server.config.OServerParameterConfiguration;
+import com.orientechnologies.orient.server.hazelcast.OHazelcastDistributedMap;
+import com.orientechnologies.orient.server.hazelcast.OHazelcastMergeStrategy;
+import com.orientechnologies.orient.server.hazelcast.OHazelcastPlugin;
+import com.orientechnologies.orient.server.network.OServerNetworkListener;
+
+/**
+ * Custom Hazelcast plugin for OrientDB which supports injecting a previously instantiated hazelcast instance.
+ */
+public class MeshOHazelcastPlugin extends OHazelcastPlugin {
+
+	public static HazelcastInstance hazelcast;
+	public static Config hazelcastConfig;
+
+	@Override
+	public HazelcastInstance configureHazelcast() throws FileNotFoundException {
+		super.hazelcastInstance = hazelcast;
+		super.hazelcastConfig = hazelcastConfig;
+		return hazelcast;
+	}
+
+	public static HazelcastInstance createHazelcast(OServerParameterConfiguration[] iParams) throws FileNotFoundException {
+		if (hazelcast == null) {
+			String hazelcastConfigFile = null;
+			for (OServerParameterConfiguration param : iParams) {
+				if (param.name.equalsIgnoreCase("configuration.hazelcast")) {
+					hazelcastConfigFile = OSystemVariableResolver.resolveSystemVariables(param.value);
+					hazelcastConfigFile = OFileUtils.getPath(hazelcastConfigFile);
+				}
+			}
+
+			if (hazelcastConfigFile == null) {
+				throw new RuntimeException("Could not determine configuration setting.");
+			}
+
+			hazelcastConfig = new FileSystemXmlConfig(hazelcastConfigFile);
+			hazelcastConfig.setClassLoader(MeshOHazelcastPlugin.class.getClassLoader());
+			hazelcastConfig.getMapConfig(CONFIG_REGISTEREDNODES).setBackupCount(6);
+			hazelcastConfig.getMapConfig(OHazelcastDistributedMap.ORIENTDB_MAP).setMergePolicy(OHazelcastMergeStrategy.class.getName());
+			hazelcastConfig.setProperty("hazelcast.shutdownhook.enabled", "false");
+			hazelcast = Hazelcast.newHazelcastInstance(hazelcastConfig);
+		}
+		return hazelcast;
+	}
+
+	@Override
+	public void shutdown() {
```

Comment:
* This is the ugly part of inheritance. This should not be overridden if it can be avoided. Dangerous hack. I think it's better to inherit from `ODistributedAbstractPlugin` instead and use the `OHazelcastPlugin` as a proxy.
* The original class calls `super.shutdown()`. This is never done here. This is very dangerous because a lot of things happen in that shutdown.

Reply:
If I inherit from `ODistributedAbstractPlugin` I will loose the `startup()` implementation and I don't want to copy that code. Yes `super.shutdown()` is never called. We can only call it if we inherit from ODistributedAbstractPlugin and duplicate the whole OHazelcast plugin code.